core/src/test/java/kafka/server/AssignmentsManagerTest.java

```diff
@@ -0,0 +1,234 @@
+                        new AssignReplicasToDirsRequestData.DirectoryData()
+                                .setId(DIR_1)
+                                .setTopics(Collections.singletonList(
+                                        new AssignReplicasToDirsRequestData.TopicData()
+                                                .setTopicId(TOPIC_1)
+                                                .setPartitions(Arrays.asList(
+                                                        new AssignReplicasToDirsRequestData.PartitionData()
+                                                                .setPartitionIndex(4),
+                                                        new AssignReplicasToDirsRequestData.PartitionData()
+                                                                .setPartitionIndex(1)
+                                                ))
+                                ))
+                ));
+        assertEquals(expected, built);
+    }
+
+    @Test
+    public void testAssignmentAggregation() throws InterruptedException {
+        CountDownLatch readyToAssert = new CountDownLatch(1);
+        doAnswer(invocation -> {
+            if (readyToAssert.getCount() > 0) {
+                readyToAssert.countDown();
+            }
+            return null;
+        }).when(channelManager).sendRequest(any(AssignReplicasToDirsRequest.Builder.class), any(ControllerRequestCompletionHandler.class));
+
+        manager.onAssignment(new TopicIdPartition(TOPIC_1, 1), DIR_1);
+        manager.onAssignment(new TopicIdPartition(TOPIC_1, 2), DIR_2);
+        manager.onAssignment(new TopicIdPartition(TOPIC_1, 3), DIR_3);
+        manager.onAssignment(new TopicIdPartition(TOPIC_1, 4), DIR_1);
+        manager.onAssignment(new TopicIdPartition(TOPIC_2, 5), DIR_2);
+        while (!readyToAssert.await(1, TimeUnit.MILLISECONDS)) {
+            time.sleep(100);
+            manager.wakeup();
+        }
+
+        ArgumentCaptor<AssignReplicasToDirsRequest.Builder> captor = ArgumentCaptor.forClass(AssignReplicasToDirsRequest.Builder.class);
+        verify(channelManager).sendRequest(captor.capture(), any(ControllerRequestCompletionHandler.class));
+        verifyNoMoreInteractions(channelManager);
+        assertEquals(1, captor.getAllValues().size());
+        AssignReplicasToDirsRequestData actual = captor.getValue().build().data();
+        AssignReplicasToDirsRequestData expected = AssignmentsManager.buildRequestData(
+                8, 100L, new HashMap<TopicIdPartition, Uuid>() {{
+                        put(new TopicIdPartition(TOPIC_1, 1), DIR_1);
+                        put(new TopicIdPartition(TOPIC_1, 2), DIR_2);
+                        put(new TopicIdPartition(TOPIC_1, 3), DIR_3);
+                        put(new TopicIdPartition(TOPIC_1, 4), DIR_1);
+                        put(new TopicIdPartition(TOPIC_2, 5), DIR_2);
+                    }}
+        );
+        assertEquals(expected, actual);
```

Comment:
This assertion is failing due to a different order of `DirectoryData.topics` here https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14369/11/testReport/junit/kafka.server/AssignmentsManagerTest/Build___JDK_8_and_Scala_2_12___testAssignmentAggregation__/